File: 198. House Robber/main.cpp
#include <vector>

using std::vector;

class Solution {
public:
    int rob(vector<int> &nums) {
        size_t size = nums.size();
        if (size == 1) {
            return nums[0];
        }

        vector<int> dp(size, 0);
        dp[0] = nums[0];
        dp[1] = std::max(dp[0], nums[1]);
        for (size_t i = 2; i < size; i++) {
            dp[i] = std::max(dp[i - 1], dp[i - 2] + nums[i]);
        }

        return dp[size - 1];
    }
};

class Solution {
public:
    int rob(vector<int> &nums) {
        int a = 0;
        int b = 0;

        for (size_t i = 0; i < nums.size(); i++) {
            if (i % 2 == 0) {
                a = std::max(a + nums[i], b);
            } else {
                b = std::max(a, b + nums[i]);
            }
        }

        return std::max(a, b);
    }
};

File: 198. House Robber/main.rs
struct Solution;

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let size = nums.len();
        if size == 1 {
            return nums[0];
        }

        let mut dp = vec![0; size];
        dp[0] = nums[0];
        dp[1] = dp[0].max(nums[1]);
        for i in 2..size {
            dp[i] = dp[i - 1].max(dp[i - 2] + nums[i]);
        }

        dp[size - 1]
    }
}

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let mut a = 0;
        let mut b = 0;

        for i in 0..nums.len() {
            if i % 2 == 0 {
                a = b.max(a + nums[i]);
            } else {
                b = a.max(b + nums[i]);
            }
        }

        a.max(b)
    }
}
